//---------------------------------------------------------------------
// Copyright 2020 National Instruments
//
// Licensed under the MIT License;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://mit-license.org/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//---------------------------------------------------------------------

//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option java_multiple_files = true;
option java_package = "labview.measurementservice";
option java_outer_classname = "MeasurementService";
option objc_class_prefix = "LVMS";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package measurementservice;

//---------------------------------------------------------------------
// The MeasurementService service definition.
//---------------------------------------------------------------------
service MeasurementService {
  
  rpc Invoke (InvokeRequest) returns (InvokeResponse) {}
  
  rpc Query (QueryRequest) returns (QueryResponse) {}
  
  rpc Register (RegistrationRequest) returns (stream ServerEvent) {}
  
  rpc OpenDMMSession (DMMOpenSessionRequest) returns (DMMOpenSessionResponse) {}
  
  rpc ConfigureDMM (DMMConfigRequest) returns (DMMConfigResponse) {}
  
  rpc PerformDMMVoltageMeasurement (DMMVoltageRequest) returns (DMMVoltageData) {}
  
  rpc CloseDMMSession (DMMCloseSessionRequest) returns (DMMCloseSessionResponse) {}
  
  rpc SelfCalibrateDMM (DMMSelfCalRequest) returns (DMMSelfCalResponse) {}
  
  rpc ResetDMM (DMMResetRequest) returns (DMMResetResponse) {}
  
  rpc OpenSMUSession (SMUOpenSessionRequest) returns (SMUOpenSessionResponse) {}
  
  rpc ConfigureSMU (SMUConfigRequest) returns (SMUConfigResponse) {}
  
  rpc EnableSMUOutput (SMUEnableRequest) returns (SMUEnableResponse) {}
  
  rpc DisableSMUOutput (SMUDisableRequest) returns (SMUDisableResponse) {}
  
  rpc PerformSMUCurrentMeasurement (SMUCurrentRequest) returns (SMUCurrentData) {}  
  
  rpc CloseSMUSession (SMUCloseSessionRequest) returns (SMUCloseSessionResponse) {}
  
  rpc SelfCalibrateSMU (SMUSelfCalRequest) returns (SMUSelfCalResponse) {}
  
  rpc ResetSMU (SMUResetRequest) returns (SMUResetResponse) {}
  
  rpc OpenNI2576MUXSession (NI2576MUXOpenSessionRequest) returns (NI2576MUXOpenSessionResponse) {}
  
  rpc CloseNI2576MUXChannel (NI2576MUXCloseChannelRequest) returns (NI2576MUXCloseChannelResponse) {}
  
  rpc OpenNI2576MUXChannel (NI2576MUXOpenChannelRequest) returns (NI2576MUXOpenChannelResponse) {}
  
  rpc DisconnectAllNI2576MUXChannels (NI2576MUXDisconnectAllRequest) returns (NI2576MUXDisconnectAllResponse) {}
  
  rpc CloseNI2576MUXSession (NI2576MUXCloseSessionRequest) returns (NI2576MUXCloseSessionResponse) {}
  
  rpc ResetNI2576MUX (NI2576MUXResetRequest) returns (NI2576MUXResetResponse) {}
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ErrorOut {
  int32 errCode = 1;
  string errMessage = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMOpenSessionRequest {
  string DMMName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMOpenSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMConfigRequest {
  int32 DMMFunction = 1;
  float DMMResolution = 2;
  float DMMRange = 3;
  int32 DMMSampleCount = 4;
  int32 DMMApertureTimeUnit = 5;
  float DMMApertureTime = 6;
  int32 DMMNumberofAverages = 7;
  int32 DMMAutoZero = 8;
  int32 DMMADCCalibration = 9;
  float DMMSettleTime = 10;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMConfigResponse {
  int32 DMMFunction = 1;
  float DMMResolution = 2;
  float DMMRange = 3;
  int32 DMMSampleCount = 4;
  int32 DMMApertureTimeUnit = 5;
  float DMMApertureTime = 6;
  int32 DMMNumberofAverages = 7;
  int32 DMMAutoZero = 8;
  int32 DMMADCCalibration = 9;
  float DMMSettleTime = 10;
  ErrorOut error = 11;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMVoltageRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMVoltageData {
  repeated float voltage = 1;
  ErrorOut error = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMCloseSessionRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMCloseSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMSelfCalRequest {
  string DMMName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMSelfCalResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMResetRequest {
  string DMMName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMResetResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUOpenSessionRequest {
  string SMUName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUOpenSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUConfigRequest {
  string SMUChannels = 1;
  int32 SMUSourceMode = 2;
  int32 SMUOutputFunction = 3;
  int32 SMUSourceTransientResponse = 4;
  float SMUCurrent = 5;
  float SMUCurrentLevelRange = 6;
  int32 SMUMeasurementSense = 7;
  int32 SMUApertureTimeUnit = 8;
  float SMUMeasurementApertureTime = 9;
  float SMUSourceAdvancedSourceDelay = 10;
  float SMUVoltage = 11;
  int32 SMUMeasurementAdvancedDCNoiseRejection = 12;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUConfigResponse {
  string SMUChannels = 1;
  int32 SMUSourceMode = 2;
  int32 SMUOutputFunction = 3;
  int32 SMUSourceTransientResponse = 4;
  float SMUCurrent = 5;
  float SMUCurrentLevelRange = 6;
  int32 SMUMeasurementSense = 7;
  int32 SMUApertureTimeUnit = 8;
  float SMUMeasurementApertureTime = 9;
  float SMUSourceAdvancedSourceDelay = 10;
  float SMUVoltage = 11;
  int32 SMUMeasurementAdvancedDCNoiseRejection = 12;
  ErrorOut error = 13;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUEnableRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUEnableResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUDisableRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUDisableResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUCurrentRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUCurrentData {
  float current = 1;
  ErrorOut error = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUCloseSessionRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUCloseSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUSelfCalRequest {
  string SMUName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUSelfCalResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUResetRequest {
  string SMUName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message SMUResetResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXOpenSessionRequest {
  string NI2576MUXName = 1;
  string NI2576MUXTopology = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXOpenSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXCloseChannelRequest {
  string Channel1 = 1;
  string Channel2 = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXCloseChannelResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXOpenChannelRequest {
  string Channel1 = 1;
  string Channel2 = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXOpenChannelResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXDisconnectAllRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXDisconnectAllResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXCloseSessionRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXCloseSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXResetRequest {
  string NI2576MUXName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message NI2576MUXResetResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeRequest {
  string command = 1;
  string parameter = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeResponse {
  int32 status = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryRequest {
  string query = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryResponse {
  string message = 1;
  int32 status = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message RegistrationRequest {
  string eventName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ServerEvent {
  string eventData = 1;
  int32 serverId = 2;
  int32 status = 3;
}
